add bot script to notify PR reviewers via slack; fix #4

martinklepsch opened the pull request:
via #4

#### Noteworthy things:

- I didn't use the `probot-slack-status` library because it seemingly didn't expose the WebClient API which I believe is needed to open new IM channels/retrieve DM channel IDs.
- For some reason `unfurl_links` does not seem to have any effect but I only tested with one PR so it might be because the link has already been seen/unfurled in recent history.
- I added a note about [Octobox](https://octobox.io/) which I think is pretty neat for managing notifications in general. 

On a side note: It might be worth copying the code of `probot-status-slack` into this repo instead of maintaining it as a separate dependency. It seems to be a relatively small library with a single 65loc file: https://github.com/PombeirP/probot-slack/blob/master/index.js

martinklepsch commented on index.js:
```diff
@@ -33,3 +34,7 @@ function getSlackMentionFromGitHubId (gitHubId) {
+
+function getSlackIdFromGitHubId (gitHubId) {
+  return SlackGitHubCacheBuilder.getSlackIdFromGitHubId(gitHubId, slackGitHubCache)
+}
```
I think it would be nice if the cache was an object exposing an interface to do conversions. Passing around functions as done with these two can become a bit troubling when functions are renamed or removed. Labels get out of sync. If they were object methods IDEs could detect + warn if they get removed or renamed (I think).

PombeirP replied:
Yeah, I think it makes sense to eventually create a service interface for this.

PombeirP replied:
Agreed.

PombeirP commented on bot_scripts/notify-reviewers-via-slack.js:
```diff
@@ -0,0 +1,49 @@
+async function notifyReviewers (context, robot, getSlackIdFromGitHubId) {
+  const { payload } = context
+
+  for (var reviewer of payload.pull_request.requested_reviewers) {
```
Why not use `let` instead of `var` by default?

PombeirP commented on bot_scripts/notify-reviewers-via-slack.js:
```diff
@@ -0,0 +1,49 @@
+    const userID = getSlackIdFromGitHubId(reviewer.login)
+
+    if (userID === undefined) {
+      robot.log.error('Could not find Slack ID for GitHub user', reviewer.login)
```
Maybe `warn` is enough here.

PombeirP commented on bot_scripts/notify-reviewers-via-slack.js:
```diff
@@ -0,0 +1,49 @@
+        robot.log.info(`${botName} - Opened DM Channel ${dmChannelID}`)
+        robot.log.info(`Notifying ${userID} about review request in ${payload.pull_request.url}`)
+
+        slackWeb.chat.postMessage(dmChannelID, msg, {unfurl_links: true, as_user: 'probot'})
```
Maybe we can have the bot user name located in a central place like `lib/slack.js`, so that future scripts reuse that instead of having it spread around?

PombeirP commented on bot_scripts/notify-reviewers-via-slack.js:
```diff
@@ -0,0 +1,49 @@
+
+const { WebClient } = require('@slack/client')
+const slackWeb = new WebClient(process.env.SLACK_BOT_TOKEN)
+const botName = 'notify-reviwers-via-slack'
```
typo in `reviewers`

PombeirP commented:
@martinklepsch Regarding `probot-status-slack`, my concern is that we'll lose future improvements to this library (which seems to be very young still). If we see it is not being maintained, that's another story.

martinklepsch commented:
[probot-slack](https://github.com/grrowl/probot-slack) has had five commits by a single contributor, the last in November. While you are right that we might miss improvements there it seems unlikely that much will happen around this library. By integrating it into our code tree (assuming LICENSE permits) we can improve it at our pace — which seems much higher than the pace of development of that library. Generally I believe less dependencies are a good thing...

martinklepsch commented:
@PombeirP addressed your feedback except the point about moving the bot name into another module. I'm not too familiar with JS (as you can see by me using `var` instead of `let`) so I'll leave this to someone who has a better understanding of JS module systems, constants naming and that kind of stuff. Would that be ok with you?

PombeirP commented:
@martinklepsch I made a live test with Adam and it worked as expected, good job! He was surprised by the octobox message, and I do think there must be a better way to inform people of alternate tools without including this every single time.

martinklepsch commented:
Hm. While I understand the surprise I think that sprinkling bot messages with helpful tips may be a good way to have information/tooling advice spread throughout the organization. Experimenting with approaches like this does seem reasonable so I'm not sure it should not even be tried?

PombeirP commented:
Octobox was already raised in Slack, so people are already aware of that possibility. I think it would be OK if we could guarantee that a user would see that only for the first time, not for every single PR.
